### ISSUE TITLE
Official Docker container images and automation to build them

### ISSUE THREAD

**Issue by vyadh** [nushell/nushell#14171]
### Related problem

I would like to use Nushell within CI/CD pipelines and other pieces of automation, but it appears there is no official image I can use. I would also like to encourage the general adoption of Nushell by helping to ensure official images are always available with each release.

### Describe the solution you'd like

A set of container images for each official release published to Docker Hub for a small set of common distros. Ideally Ubuntu for my purposes, but Debian and Alpine seem more generally useful.

I'd like to either take the existing images and automation by @hustcer (Hi!) and help get them promoted as official Nushell images along with any tests to gain confidence that what is produced is working. Or, if that's not possible for some reason, start from scratch building off the existing `latest` image.

### Describe alternatives you've considered

There are [the set of images](https://hub.docker.com/r/hustcer/nushell) provided by @hustcer on Docker Hub that are regularly updated with new releases. However, while @hustcer appears to be a Nushell contributor, many companies would struggle to adopt container images that do not have official project backing and review. I also can't find the source code repo that generates those images, which seems important to have for any images you want to rely on.

There is also the existing Dockerfile but would need to build it myself and it's only provided for `latest`.

Alternatively, I could build a set of images and automation for my own purposes but it seems a shame to duplicate effort.

### Additional context and details

To be clear, I'm happy doing this work. I mainly wanted to get some validation that this is desirable by Nushell's maintainers before I put any work in.

**Comment by hustcer** [comment]
Thanks for the suggestion, I like this idea very much, and it's true that I have already created a set of workflows to build the `nushell` image regularly, they are built from the official release binaries. due to a few reasons they are not included in the nushell repo:
1. There are enough workflows now in the nushell repo, I'm hesitate to add more
2. It seems that nushell don't have an official account on docker hub, alternatively, the images could be pushed to `ghcr.io`
3. No one has brought this up before, so I assumed the need wasn't very strong

Anyway I can add these to the nushell repository if that's ok with everyone.

**Comment by vyadh** [comment]
Thanks for the reply and for what you already provide. I for one would love to have those migrated as the official images. I suppose if the images are in this repo it's a bit easier and automatic to trigger a new set of images as an official release happens. I'm more than willing to add any tests around the image build if any are deemed needed.

**Comment by hustcer** [comment]
@fdncred How about adding a workflow to build alpine images for each Nushell release?

**Comment by fdncred** [comment]
What would that look like? I do my best to not use docker at all so I'm somewhat ignorant.

**Comment by hustcer** [comment]
> What would that look like? I do my best to not use docker at all so I'm somewhat ignorant.

Simply put, `Nushell` will be installed into a lightweight container, our users just pull down the container as easy as `git pull` then they can run it immediately without building it.

**Comment by fdncred** [comment]
yes, i understand the purpose of docker but i don't want to clutter up the repo. This is why we have an integrations repo. what i was asking about is what would it look like in the repo and what would the ci look like.

**Comment by hustcer** [comment]
> yes, i understand the purpose of docker but i don't want to clutter up the repo. This is why we have an integrations repo. what i was asking about is what would it look like in the repo and what would the ci look like.

We will add one workflow to build the `nushell` alpine image, the workflow will run after each release published(almost once monthly), that's to say It won't block the release(even if it fails), It's better to put it in the nushell/nushell repo rather than integration repo.

**Comment by vyadh** [comment]
Maybe you'd be open to me adding a Debian and/or Ubuntu version in addition to Alpine? Bigger image, but a bit more compatible with Debian and APT-based toolchains. I'm sure we could fit that into the same workflow via a matrix build. We can also try to make the Alpine and Debian Dockerfile as similar as possible to ease maintenance.

Depending on how sophisticated we want to get, I suppose we might want to rebuild the latest version more than monthly to address any vulnerabilities in the base image, perhaps the same workflow could be scheduled nightly and do a rebuild if the upstream image has changed. This is more of a nice to have though. Something to come back to after the initial implementation perhaps.

**Comment by fdncred** [comment]
I'm not opposed to another CI as long as it doesn't get in the way, and someone maintains it. My preference would for it to be in the integrations repo though.

**Comment by hustcer** [comment]
I have a better idea: we can keep the docker files in nushell repo and workflow in nightly repo,so we can build images for each release along with a nightly tag.

**Comment by fdncred** [comment]
ya, i'm fine with it being in the nightly repo.

**Comment by hustcer** [comment]
https://github.com/nushell/nightly/pkgs/container/nushell

```shell
docker pull ghcr.io/nushell/nushell:latest
```

**Comment by vyadh** [comment]
> https://github.com/nushell/nightly/pkgs/container/nushell
> 
> ```shell
> docker pull ghcr.io/nushell/nushell:latest
> ```

Fantastic. Thank you. I'll add the Debian/Ubuntu version to that if we're still happy to officially support both.
